### DISCUSSION TITLE
Strange DPY-5002 error with executemany batcherrors

### DISCUSSION THREAD

**Discussion by vinidmpereira** [oracle/python-oracledb#261] [Q&A] [answered]
Hello all,

I'm currently having a weird problem in Airflow using OracleHook, i know this forum pertains to oracledb but i think this is where i need help.

So i've been writing a large migrating script  to unify multiple similar databases into a singular one, and in one of the migrations i've been hit with:

 `oracledb.exceptions.InternalError: DPY-5002: internal error: read integer of length 3 when expecting integer of no more than length 2`

And this error only occurs in one of the 40+ databases i'm currently migrating, it's such a weird error and i can't find anything about this specific error, theres another topic where someone is currently having a similar mistake but with a DPY-5003 error, and i don't know if they are connected.

**Comment by anthony-tuininga** [comment]
Both of those errors indicate that the server is sending something that the client doesn't know how to process properly. If you can send packet traces that would be most helpful (set the environment variable `PYO_DEBUG_PACKETS` to any value before running your script). It would also be helpful to know the version of the driver you are using, the version of the database you are connecting to, the platform you are using and so forth. The full traceback is also very useful to know *where* the client is running into trouble!

**Reply by vinidmpereira** [reply]
My environment is kinda ... fuzzy, but i'll get this data and get back to you as soon as i can.

**Reply by anthony-tuininga** [reply]
The full traceback is immediately useful (you can trim any lines in your code if you wish). The packet trace (of the last packet sent to the server and the response from the server) should make it possible to diagnose the problem. The other information may be of help but can come later if needed!

**Reply by cjbj** [reply]
@vinidmpereira If you don't want to share it on GH, mail us or open an Oracle Support ticket and attach it there.

**Reply by vinidmpereira** [reply]
It's really not that, the environment is really awkward to "reason" with, i'm trying to reproduce it on localhost and it's not triggering, whatever it is.

On our DEV environment it's the same, and i'll only be able to hit HML later tonight (where the error happened the last time). i''l provide the traceback as soon as i'm able.

I'm not sure i'll be able to packet trace on our HML environment, i have very little power on that one to set environment variables.

**Reply by cjbj** [reply]
What versions of the db etc are you using?  What OS?

**Reply by vinidmpereira** [reply]
Localhost and DEV and HML environments 
OS is a Docker Container providedo by Airflow
"Debian GNU/Linux 11 (bullseye)"

Both DEV and HML are a standar helm image deployed on kubernetes.

DEV DB
Oracle Database 19c Enterprise Edition Release 19.0.0.0.0
HML DB
Oracle Database 19c Enterprise Edition Release 19.0.0.0.0 

oracledb - 1.3.2 on both environments.


Traceback was slightly edited and some of timestamped data was removed mainly to help with readability.
Traceback gist:
https://gist.github.com/vinidmpereira/09ac86d6d8ef81039f08d5caca947406

Source code is a modified OracleHook bulkwrite.
Source Code gist:
https://gist.github.com/vinidmpereira/c69a01007778329b700ae82fd2cb6a12

**Reply by cjbj** [reply]
All info is good info - thanks.

What are the actual DB versions?  For reasons we never liked, the DB team return only the major release in some places but the actual version will have a significant second number.  E.g. 19.21.  Can you share what `Connection.version` shows?

**Reply by vinidmpereira** [reply]
Both DB's are 19.16.1.0.0

**Reply by anthony-tuininga** [reply]
How many rows are you inserting? And how many batch errors are you expecting to be returned? The location in the code suggests an issue with this -- I will see if I can replicate based on what I am seeing.

**Reply by vinidmpereira** [reply]
About 250k per round, the data is sliced and then pass trough that code.

Error count varies, worst case about 10k, best scenario no error whatsoever.

**Reply by vinidmpereira** [reply]
And now that you've said something about quantities, i've fixed this before, by slicing datasets i believe.

**Reply by anthony-tuininga** [reply, marked as answer]
Ok. I can replicate. The protocol only allows for 65535 errors. If you exceed this value you get this error in thin mode. In thick mode you get invalid results (the number of batch errors is simply truncated by looking at the lowest 16 bits of the value). As such, this is a bug in the database and will need to be fixed there. In the meantime, you can avoid the issue by always ensuring that you never send more than 65535 rows to the database at a time when using batch errors.

**Reply by vinidmpereira** [reply]
You have no idea how happy knowing that it's easily avoidableand to know what causes this, i'll lower my dataset size on processing and continue testing, thanks a lot.

**Reply by cjbj** [reply]
Or lower your bad data rate :)

**Reply by vinidmpereira** [reply]
Believe me, i would if i could, but my hands are tied.

**Reply by cjbj** [reply]
python-oracledb 2.0.0 had improvements in messages when a large number of errors occur.
